File: internal/services/redshift/dataapi.go
package redshift

import (
	"bytes"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"
)

type statement struct {
	ID                string
	ClusterIdentifier string
	Database          string
	DbUser            string
	SessionID         string
	QueryString       string
	ResultFormat      string
	CreatedAt         time.Time
	UpdatedAt         time.Time
	Status            string
	Error             string
	HasResultSet      bool
	Result            queryResult
}

type session struct {
	ID                      string
	CreatedAt               time.Time
	UpdatedAt               time.Time
	ExpiresAt               time.Time
	SessionKeepAliveSeconds int
}

type executeStatementRequest struct {
	ClusterIdentifier       string
	Database                string
	DbUser                  string
	SQL                     string `json:"Sql"`
	ClientToken             string
	SessionID               string `json:"SessionId"`
	SessionKeepAliveSeconds int
	ResultFormat            string
}

type batchExecuteStatementRequest struct {
	ClusterIdentifier       string
	Database                string
	DbUser                  string
	SQLs                    []string `json:"Sqls"`
	ClientToken             string
	SessionID               string `json:"SessionId"`
	SessionKeepAliveSeconds int
	ResultFormat            string
}

type statementIDRequest struct {
	ID string `json:"Id"`
}

type getStatementResultRequest struct {
	ID         string `json:"Id"`
	MaxResults int
	NextToken  string
}

type listStatementsRequest struct {
	Status string
}

type listMetadataRequest struct {
	ClusterIdentifier string
	ConnectedDatabase string
	Database          string
	DbUser            string
	Schema            string
	SchemaPattern     string
	TablePattern      string
	MaxResults        int
	NextToken         string
}

type describeTableRequest struct {
	ClusterIdentifier string
	ConnectedDatabase string
	Database          string
	DbUser            string
	Schema            string
	Table             string
	MaxResults        int
	NextToken         string
}

type executeStatementResponse struct {
	ID                string `json:"Id"`
	ClusterIdentifier string
	Database          string
	DbUser            string
	SessionID         string `json:"SessionId,omitempty"`
	CreatedAt         int64
}

type describeStatementResponse struct {
	ID                string `json:"Id"`
	ClusterIdentifier string
	Database          string
	DbUser            string
	SessionID         string `json:"SessionId,omitempty"`
	QueryString       string
	Status            string
	Error             string `json:",omitempty"`
	CreatedAt         int64
	UpdatedAt         int64
	Duration          int64
	HasResultSet      bool
	ResultRows        int64
	ResultSize        int64
	RedshiftQueryID   int64 `json:"RedshiftQueryId"`
}

type statementListItem struct {
	ID           string `json:"Id"`
	QueryString  string
	Status       string
	CreatedAt    int64
	UpdatedAt    int64
	HasResultSet bool
}

type dataAPIResultField struct {
	IsNull       bool     `json:"isNull,omitempty"`
	LongValue    *int64   `json:"longValue,omitempty"`
	DoubleValue  *float64 `json:"doubleValue,omitempty"`
	BooleanValue *bool    `json:"booleanValue,omitempty"`
	StringValue  *string  `json:"stringValue,omitempty"`
}

type columnMetadata struct {
	Name            string `json:"name"`
	Label           string `json:"label"`
	SchemaName      string `json:"schemaName,omitempty"`
	TableName       string `json:"tableName,omitempty"`
	TypeName        string `json:"typeName"`
	ColumnDefault   string `json:"columnDefault,omitempty"`
	IsCaseSensitive bool   `json:"isCaseSensitive"`
	IsSigned        bool   `json:"isSigned"`
	Nullable        int    `json:"nullable"`
	Precision       int    `json:"precision"`
	Scale           int    `json:"scale"`
}

type tableMember struct {
	Name   string
	Schema string
	Type   string
}

func decodeDataAPIRequest(w http.ResponseWriter, r *http.Request, target any) bool {
	defer r.Body.Close()
	if err := json.NewDecoder(r.Body).Decode(target); err != nil {
		writeDataAPIError(w, http.StatusBadRequest, "ValidationException", "invalid JSON request")
		return false
	}
	return true
}

func (s *Server) nextStatementIDValue() string {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.nextStatementID++
	return fmt.Sprintf("devcloud-redshift-%d", s.nextStatementID)
}

func (s *Server) statementByID(w http.ResponseWriter, id string) *statement {
	if id == "" {
		writeDataAPIError(w, http.StatusBadRequest, "ValidationException", "Id is required")
		return nil
	}
	s.mu.Lock()
	stmt := s.statements[id]
	s.mu.Unlock()
	if stmt == nil {
		writeDataAPIError(w, http.StatusNotFound, "ResourceNotFoundException", "statement does not exist")
		return nil
	}
	return stmt
}

func executeStatementResponseFromStatement(stmt *statement) executeStatementResponse {
	return executeStatementResponse{
		ID:                stmt.ID,
		ClusterIdentifier: stmt.ClusterIdentifier,
		Database:          stmt.Database,
		DbUser:            stmt.DbUser,
		SessionID:         stmt.SessionID,
		CreatedAt:         stmt.CreatedAt.Unix(),
	}
}

func describeStatementResponseFromStatement(stmt *statement) describeStatementResponse {
	return describeStatementResponse{
		ID:                stmt.ID,
		ClusterIdentifier: stmt.ClusterIdentifier,
		Database:          stmt.Database,
		DbUser:            stmt.DbUser,
		SessionID:         stmt.SessionID,
		QueryString:       safeStatementQueryString(stmt.QueryString),
		Status:            stmt.Status,
		Error:             stmt.Error,
		CreatedAt:         stmt.CreatedAt.Unix(),
		UpdatedAt:         stmt.UpdatedAt.Unix(),
		HasResultSet:      stmt.HasResultSet,
		ResultRows:        int64(len(stmt.Result.rows)),
		ResultSize:        approximateResultSize(stmt.Result),
		RedshiftQueryID:   redshiftQueryID(stmt.ID),
	}
}

func getStatementResultResponse(result queryResult, rows [][]string) map[string]any {
	records := make([][]dataAPIResultField, 0, len(rows))
	for _, row := range rows {
		fields := make([]dataAPIResultField, 0, len(row))
		for i, value := range row {
			typeOID := pgTypeVarcharOID
			if i < len(result.fields) {
				typeOID = result.fields[i].TypeOID
			}
			fields = append(fields, dataAPIField(value, typeOID))
		}
		records = append(records, fields)
	}
	metadata := make([]columnMetadata, 0, len(result.fields))
	for i, field := range result.fields {
		metadata = append(metadata, columnMetadataFromPGField(field, i))
	}
	return map[string]any{
		"ColumnMetadata": metadata,
		"Records":        records,
		"TotalNumRows":   len(result.rows),
	}
}

func getStatementResultV2Response(result queryResult, rows [][]string) (map[string]any, error) {
	records := make([]map[string]string, 0, len(rows))
	for _, row := range rows {
		record, err := csvRecord(row)
		if err != nil {
			return nil, err
		}
		records = append(records, map[string]string{"CSVRecords": record})
	}
	metadata := make([]columnMetadata, 0, len(result.fields))
	for i, field := range result.fields {
		metadata = append(metadata, columnMetadataFromPGField(field, i))
	}
	return map[string]any{
		"ColumnMetadata": metadata,
		"Records":        records,
		"ResultFormat":   "CSV",
		"TotalNumRows":   len(result.rows),
	}, nil
}

func csvRecord(row []string) (string, error) {
	var buf bytes.Buffer
	writer := csv.NewWriter(&buf)
	if err := writer.Write(row); err != nil {
		return "", err
	}
	writer.Flush()
	if err := writer.Error(); err != nil {
		return "", err
	}
	return strings.TrimSuffix(buf.String(), "\n"), nil
}

func dataAPIField(value string, typeOID int32) dataAPIResultField {
	switch typeOID {
	case pgTypeInt4OID:
		if parsed, err := strconv.ParseInt(value, 10, 64); err == nil {
			return dataAPIResultField{LongValue: &parsed}
		}
	case pgTypeFloat8OID:
		if parsed, err := strconv.ParseFloat(value, 64); err == nil {
			return dataAPIResultField{DoubleValue: &parsed}
		}
	case pgTypeBoolOID:
		if parsed, err := strconv.ParseBool(strings.ToLower(value)); err == nil {
			return dataAPIResultField{BooleanValue: &parsed}
		}
	}
	return dataAPIResultField{StringValue: &value}
}

func columnMetadataFromPGField(field pgField, ordinal int) columnMetadata {
	typeName := "varchar"
	precision := 256
	signed := false
	switch field.TypeOID {
	case pgTypeInt4OID:
		typeName = "int4"
		precision = 10
		signed = true
	case pgTypeFloat8OID:
		typeName = "float8"
		precision = 17
		signed = true
	case pgTypeBoolOID:
		typeName = "bool"
		precision = 1
	}
	return columnMetadata{
		Name:            field.Name,
		Label:           field.Name,
		TypeName:        typeName,
		IsCaseSensitive: false,
		IsSigned:        signed,
		Nullable:        1,
		Precision:       precision,
		Scale:           0,
	}
}

func columnMetadataFromColumn(column column, ordinal int) columnMetadata {
	return columnMetadataFromPGField(pgField{
		Name:     column.name,
		TypeOID:  pgTypeOID(column.dataType),
		TypeSize: pgTypeSize(column.dataType),
	}, ordinal)
}

func approximateResultSize(result queryResult) int64 {
	var size int64
	for _, row := range result.rows {
		for _, value := range row {
			size += int64(len(value))
		}
	}
	return size
}

func redshiftQueryID(id string) int64 {
	var result int64
	for _, ch := range id {
		result = result*31 + int64(ch)
		if result < 0 {
			result = -result
		}
	}
	if result == 0 {
		return 1
	}
	return result
}

func safeStatementQueryString(sql string) string {
	preview, redacted, _ := safeSQLPreview(sql, len(sql))
	if redacted {
		return preview
	}
	return sql
}

func (s *Server) nextSessionIDValue() string {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.nextSessionID++
	return fmt.Sprintf("devcloud-redshift-session-%d", s.nextSessionID)
}

func (s *Server) sessionIDForRequest(sessionID string, keepAliveSeconds int, now time.Time) (string, error) {
	if keepAliveSeconds < 0 {
		return "", fmt.Errorf("SessionKeepAliveSeconds must be non-negative")
	}
	sessionID = strings.TrimSpace(sessionID)
	if sessionID == "" && keepAliveSeconds <= 0 {
		return "", nil
	}
	if sessionID == "" {
		sessionID = s.nextSessionIDValue()
	}
	expiresAt := now
	if keepAliveSeconds > 0 {
		expiresAt = now.Add(time.Duration(keepAliveSeconds) * time.Second)
	}
	s.mu.Lock()
	if existing := s.sessions[sessionID]; existing != nil {
		existing.UpdatedAt = now
		existing.SessionKeepAliveSeconds = keepAliveSeconds
		existing.ExpiresAt = expiresAt
	} else {
		s.sessions[sessionID] = &session{
			ID:                      sessionID,
			CreatedAt:               now,
			UpdatedAt:               now,
			ExpiresAt:               expiresAt,
			SessionKeepAliveSeconds: keepAliveSeconds,
		}
	}
	s.mu.Unlock()
	return sessionID, nil
}

func safeSQLPreview(sql string, maxBytes int) (string, bool, bool) {
	preview := strings.Join(strings.Fields(sql), " ")
	if preview == "" || maxBytes <= 0 {
		return "", false, false
	}
	lower := strings.ToLower(preview)
	for _, token := range []string{
		"authorization",
		"credentials",
		"access_key_id",
		"secret_access_key",
		"session_token",
		"iam_role",
		"password",
	} {
		if strings.Contains(lower, token) {
			return "[redacted]", true, false
		}
	}
	if len(preview) <= maxBytes {
		return preview, false, false
	}
	return preview[:maxBytes], false, true
}

File: services/redshift/src/dataapi.rs
//! Redshift Data API request/response models and pure helpers.
//!
//! Parity: `internal/services/redshift/dataapi.rs`. The handler wiring lives in
//! `http_api.rs`; this module holds the typed AWS-JSON request/response structs
//! and the value/metadata conversion helpers (`dataAPIField`,
//! `columnMetadataFromPGField`, pagination, CSV records, redaction).

use serde::{Deserialize, Serialize};

use crate::engine::QueryResult;
use crate::errors::SqlError;
use crate::model::Column;
use crate::pg_types::{
    pg_type_oid, pg_type_size, PgField, PG_TYPE_BOOL_OID, PG_TYPE_FLOAT8_OID, PG_TYPE_INT4_OID,
    PG_TYPE_VARCHAR_OID,
};
use crate::server::StatementRecord;
use crate::snapshot::safe_sql_preview;
use crate::storage::unix_seconds;

// --- requests --------------------------------------------------------------

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct ExecuteStatementRequest {
    #[serde(rename = "ClusterIdentifier")]
    pub cluster_identifier: String,
    #[serde(rename = "Database")]
    pub database: String,
    #[serde(rename = "DbUser")]
    pub db_user: String,
    #[serde(rename = "Sql")]
    pub sql: String,
    #[serde(rename = "ClientToken")]
    pub client_token: String,
    #[serde(rename = "SessionId")]
    pub session_id: String,
    #[serde(rename = "SessionKeepAliveSeconds")]
    pub session_keep_alive_seconds: i64,
    #[serde(rename = "ResultFormat")]
    pub result_format: String,
}

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct BatchExecuteStatementRequest {
    #[serde(rename = "ClusterIdentifier")]
    pub cluster_identifier: String,
    #[serde(rename = "Database")]
    pub database: String,
    #[serde(rename = "DbUser")]
    pub db_user: String,
    #[serde(rename = "Sqls")]
    pub sqls: Vec<String>,
    #[serde(rename = "ClientToken")]
    pub client_token: String,
    #[serde(rename = "SessionId")]
    pub session_id: String,
    #[serde(rename = "SessionKeepAliveSeconds")]
    pub session_keep_alive_seconds: i64,
    #[serde(rename = "ResultFormat")]
    pub result_format: String,
}

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct StatementIdRequest {
    #[serde(rename = "Id")]
    pub id: String,
}

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct GetStatementResultRequest {
    #[serde(rename = "Id")]
    pub id: String,
    #[serde(rename = "MaxResults")]
    pub max_results: i64,
    #[serde(rename = "NextToken")]
    pub next_token: String,
}

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct ListStatementsRequest {
    #[serde(rename = "Status")]
    pub status: String,
}

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct ListMetadataRequest {
    #[serde(rename = "ClusterIdentifier")]
    pub cluster_identifier: String,
    #[serde(rename = "ConnectedDatabase")]
    pub connected_database: String,
    #[serde(rename = "Database")]
    pub database: String,
    #[serde(rename = "DbUser")]
    pub db_user: String,
    #[serde(rename = "Schema")]
    pub schema: String,
    #[serde(rename = "SchemaPattern")]
    pub schema_pattern: String,
    #[serde(rename = "TablePattern")]
    pub table_pattern: String,
    #[serde(rename = "MaxResults")]
    pub max_results: i64,
    #[serde(rename = "NextToken")]
    pub next_token: String,
}

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct DescribeTableRequest {
    #[serde(rename = "ClusterIdentifier")]
    pub cluster_identifier: String,
    #[serde(rename = "ConnectedDatabase")]
    pub connected_database: String,
    #[serde(rename = "Database")]
    pub database: String,
    #[serde(rename = "DbUser")]
    pub db_user: String,
    #[serde(rename = "Schema")]
    pub schema: String,
    #[serde(rename = "Table")]
    pub table: String,
    #[serde(rename = "MaxResults")]
    pub max_results: i64,
    #[serde(rename = "NextToken")]
    pub next_token: String,
}

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct ServerlessListRequest {
    #[serde(rename = "maxResults")]
    pub max_results: i64,
    #[serde(rename = "nextToken")]
    pub next_token: String,
}

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct ServerlessNamespaceRequest {
    #[serde(rename = "namespaceName")]
    pub namespace_name: String,
}

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct ServerlessWorkgroupRequest {
    #[serde(rename = "workgroupName")]
    pub workgroup_name: String,
}

// --- responses -------------------------------------------------------------

#[derive(Debug, Default, Serialize)]
pub struct ExecuteStatementResponse {
    #[serde(rename = "Id")]
    pub id: String,
    #[serde(rename = "ClusterIdentifier")]
    pub cluster_identifier: String,
    #[serde(rename = "Database")]
    pub database: String,
    #[serde(rename = "DbUser")]
    pub db_user: String,
    #[serde(rename = "SessionId", skip_serializing_if = "String::is_empty")]
    pub session_id: String,
    #[serde(rename = "CreatedAt")]
    pub created_at: i64,
}

#[derive(Debug, Default, Serialize)]
pub struct DescribeStatementResponse {
    #[serde(rename = "Id")]
    pub id: String,
    #[serde(rename = "ClusterIdentifier")]
    pub cluster_identifier: String,
    #[serde(rename = "Database")]
    pub database: String,
    #[serde(rename = "DbUser")]
    pub db_user: String,
    #[serde(rename = "SessionId", skip_serializing_if = "String::is_empty")]
    pub session_id: String,
    #[serde(rename = "QueryString")]
    pub query_string: String,
    #[serde(rename = "Status")]
    pub status: String,
    #[serde(rename = "Error", skip_serializing_if = "String::is_empty")]
    pub error: String,
    #[serde(rename = "CreatedAt")]
    pub created_at: i64,
    #[serde(rename = "UpdatedAt")]
    pub updated_at: i64,
    #[serde(rename = "Duration")]
    pub duration: i64,
    #[serde(rename = "HasResultSet")]
    pub has_result_set: bool,
    #[serde(rename = "ResultRows")]
    pub result_rows: i64,
    #[serde(rename = "ResultSize")]
    pub result_size: i64,
    #[serde(rename = "RedshiftQueryId")]
    pub redshift_query_id: i64,
}

#[derive(Debug, Default, Serialize)]
pub struct StatementListItem {
    #[serde(rename = "Id")]
    pub id: String,
    #[serde(rename = "QueryString")]
    pub query_string: String,
    #[serde(rename = "Status")]
    pub status: String,
    #[serde(rename = "CreatedAt")]
    pub created_at: i64,
    #[serde(rename = "UpdatedAt")]
    pub updated_at: i64,
    #[serde(rename = "HasResultSet")]
    pub has_result_set: bool,
}

/// Mirrors `dataAPIResultField`. legacy uses pointers + omitempty so exactly one of
/// the value fields (or `isNull`) is present; `Option` + skip mirrors that.
#[derive(Debug, Default, Serialize)]
pub struct DataApiResultField {
    #[serde(rename = "isNull", skip_serializing_if = "is_false")]
    pub is_null: bool,
    #[serde(rename = "longValue", skip_serializing_if = "Option::is_none")]
    pub long_value: Option<i64>,
    #[serde(rename = "doubleValue", skip_serializing_if = "Option::is_none")]
    pub double_value: Option<f64>,
    #[serde(rename = "booleanValue", skip_serializing_if = "Option::is_none")]
    pub boolean_value: Option<bool>,
    #[serde(rename = "stringValue", skip_serializing_if = "Option::is_none")]
    pub string_value: Option<String>,
}

#[derive(Debug, Default, Serialize)]
pub struct ColumnMetadata {
    pub name: String,
    pub label: String,
    #[serde(rename = "schemaName", skip_serializing_if = "String::is_empty")]
    pub schema_name: String,
    #[serde(rename = "tableName", skip_serializing_if = "String::is_empty")]
    pub table_name: String,
    #[serde(rename = "typeName")]
    pub type_name: String,
    #[serde(rename = "columnDefault", skip_serializing_if = "String::is_empty")]
    pub column_default: String,
    #[serde(rename = "isCaseSensitive")]
    pub is_case_sensitive: bool,
    #[serde(rename = "isSigned")]
    pub is_signed: bool,
    pub nullable: i64,
    pub precision: i64,
    pub scale: i64,
}

#[derive(Debug, Clone)]
pub struct TableMember {
    pub name: String,
    pub schema: String,
    pub member_type: String,
}

fn is_false(value: &bool) -> bool {
    !*value
}

// --- conversions -----------------------------------------------------------

/// Mirrors `executeStatementResponseFromStatement`.
pub fn execute_statement_response_from_statement(
    stmt: &StatementRecord,
) -> ExecuteStatementResponse {
    ExecuteStatementResponse {
        id: stmt.id.clone(),
        cluster_identifier: stmt.cluster_identifier.clone(),
        database: stmt.database.clone(),
        db_user: stmt.db_user.clone(),
        session_id: stmt.session_id.clone(),
        created_at: unix_seconds(stmt.created_at),
    }
}

/// Mirrors `describeStatementResponseFromStatement`.
pub fn describe_statement_response_from_statement(
    stmt: &StatementRecord,
) -> DescribeStatementResponse {
    DescribeStatementResponse {
        id: stmt.id.clone(),
        cluster_identifier: stmt.cluster_identifier.clone(),
        database: stmt.database.clone(),
        db_user: stmt.db_user.clone(),
        session_id: stmt.session_id.clone(),
        query_string: safe_statement_query_string(&stmt.query_string),
        status: stmt.status.clone(),
        error: stmt.error.clone(),
        created_at: unix_seconds(stmt.created_at),
        updated_at: unix_seconds(stmt.updated_at),
        duration: 0,
        has_result_set: stmt.has_result_set,
        result_rows: stmt.result.rows.len() as i64,
        result_size: approximate_result_size(&stmt.result),
        redshift_query_id: crate::snapshot::redshift_query_id(&stmt.id),
    }
}

/// Mirrors `getStatementResultResponse` as an ordered JSON object.
pub fn get_statement_result_response(
    result: &QueryResult,
    rows: &[Vec<String>],
) -> serde_json::Value {
    let records: Vec<Vec<DataApiResultField>> = rows
        .iter()
        .map(|row| {
            row.iter()
                .enumerate()
                .map(|(i, value)| {
                    let type_oid = result
                        .fields
                        .get(i)
                        .map(|f| f.type_oid)
                        .unwrap_or(PG_TYPE_VARCHAR_OID);
                    data_api_field(value, type_oid)
                })
                .collect()
        })
        .collect();
    let metadata: Vec<ColumnMetadata> = result
        .fields
        .iter()
        .enumerate()
        .map(|(i, field)| column_metadata_from_pg_field(field, i))
        .collect();
    serde_json::json!({
        "ColumnMetadata": metadata,
        "Records": records,
        "TotalNumRows": result.rows.len(),
    })
}

/// Mirrors `getStatementResultV2Response`.
pub fn get_statement_result_v2_response(
    result: &QueryResult,
    rows: &[Vec<String>],
) -> Result<serde_json::Value, SqlError> {
    let mut records = Vec::with_capacity(rows.len());
    for row in rows {
        let record = csv_record(row)?;
        records.push(serde_json::json!({ "CSVRecords": record }));
    }
    let metadata: Vec<ColumnMetadata> = result
        .fields
        .iter()
        .enumerate()
        .map(|(i, field)| column_metadata_from_pg_field(field, i))
        .collect();
    Ok(serde_json::json!({
        "ColumnMetadata": metadata,
        "Records": records,
        "ResultFormat": "CSV",
        "TotalNumRows": result.rows.len(),
    }))
}

/// Mirrors `csvRecord` (legacy encoding/csv: comma-separated, quote on need,
/// doubled inner quotes, no trailing newline).
pub fn csv_record(row: &[String]) -> Result<String, SqlError> {
    let mut out = String::new();
    for (i, field) in row.iter().enumerate() {
        if i > 0 {
            out.push(',');
        }
        if csv_field_needs_quoting(field) {
            out.push('"');
            out.push_str(&field.replace('"', "\"\""));
            out.push('"');
        } else {
            out.push_str(field);
        }
    }
    Ok(out)
}

/// legacy encoding/csv quotes a field that contains a comma, quote, CR, LF, or
/// that begins with a space, and quotes a field equal to a lone `﻿` BOM.
fn csv_field_needs_quoting(field: &str) -> bool {
    if field.is_empty() {
        return false;
    }
    if field.contains([',', '"', '\r', '\n']) {
        return true;
    }
    field.starts_with(' ')
}

/// Mirrors `dataAPIField`.
pub fn data_api_field(value: &str, type_oid: i32) -> DataApiResultField {
    match type_oid {
        PG_TYPE_INT4_OID => {
            if let Ok(parsed) = value.parse::<i64>() {
                return DataApiResultField {
                    long_value: Some(parsed),
                    ..DataApiResultField::default()
                };
            }
        }
        PG_TYPE_FLOAT8_OID => {
            if let Ok(parsed) = parse_go_float(value) {
                return DataApiResultField {
                    double_value: Some(parsed),
                    ..DataApiResultField::default()
                };
            }
        }
        PG_TYPE_BOOL_OID => {
            if let Some(parsed) = parse_go_bool(&value.to_lowercase()) {
                return DataApiResultField {
                    boolean_value: Some(parsed),
                    ..DataApiResultField::default()
                };
            }
        }
        _ => {}
    }
    DataApiResultField {
        string_value: Some(value.to_string()),
        ..DataApiResultField::default()
    }
}

/// legacy `strconv.ParseBool` accepts 1,t,T,TRUE,true,True,0,f,F,FALSE,false,False.
fn parse_go_bool(value: &str) -> Option<bool> {
    match value {
        "1" | "t" | "true" => Some(true),
        "0" | "f" | "false" => Some(false),
        _ => None,
    }
}

/// legacy `strconv.ParseFloat` (64-bit). Rust's `f64::from_str` is compatible for
/// the decimal forms devcloud emits.
fn parse_go_float(value: &str) -> Result<f64, ()> {
    value.parse::<f64>().map_err(|_| ())
}

/// Mirrors `columnMetadataFromPGField`.
pub fn column_metadata_from_pg_field(field: &PgField, _ordinal: usize) -> ColumnMetadata {
    let (type_name, precision, signed) = match field.type_oid {
        PG_TYPE_INT4_OID => ("int4", 10, true),
        PG_TYPE_FLOAT8_OID => ("float8", 17, true),
        PG_TYPE_BOOL_OID => ("bool", 1, false),
        _ => ("varchar", 256, false),
    };
    ColumnMetadata {
        name: field.name.clone(),
        label: field.name.clone(),
        type_name: type_name.to_string(),
        is_case_sensitive: false,
        is_signed: signed,
        nullable: 1,
        precision,
        scale: 0,
        ..ColumnMetadata::default()
    }
}

/// Mirrors `columnMetadataFromColumn`.
pub fn column_metadata_from_column(column: &Column, ordinal: usize) -> ColumnMetadata {
    column_metadata_from_pg_field(
        &PgField {
            name: column.name.clone(),
            type_oid: pg_type_oid(&column.data_type),
            type_size: pg_type_size(&column.data_type),
        },
        ordinal,
    )
}

/// Mirrors `approximateResultSize`.
pub fn approximate_result_size(result: &QueryResult) -> i64 {
    let mut size = 0;
    for row in &result.rows {
        for value in row {
            size += value.len() as i64;
        }
    }
    size
}

/// Mirrors `safeStatementQueryString`.
pub fn safe_statement_query_string(sql: &str) -> String {
    let (preview, redacted, _) = safe_sql_preview(sql, sql.len());
    if redacted {
        preview
    } else {
        sql.to_string()
    }
}

// --- pagination ------------------------------------------------------------

/// Mirrors `paginationStart`.
pub fn pagination_start(next_token: &str) -> Result<usize, SqlError> {
    if next_token.is_empty() {
        return Ok(0);
    }
    match next_token.parse::<i64>() {
        Ok(start) if start >= 0 => Ok(start as usize),
        _ => Err(SqlError::new("NextToken is invalid")),
    }
}

/// Mirrors `paginationEnd`.
pub fn pagination_end(start: usize, total: usize, max_results: i64) -> usize {
    if max_results <= 0 || start + (max_results as usize) > total {
        return total;
    }
    start + (max_results as usize)
}

/// Mirrors the `paginate*` family: returns (page, next_token). The page is a
/// borrowed slice of `values`.
pub fn paginate<'a, T>(
    values: &'a [T],
    max_results: i64,
    next_token: &str,
) -> Result<(&'a [T], String), SqlError> {
    let start = pagination_start(next_token)?;
    if start >= values.len() {
        return Ok((&[], String::new()));
    }
    let end = pagination_end(start, values.len(), max_results);
    let next = if end < values.len() {
        end.to_string()
    } else {
        String::new()
    };
    Ok((&values[start..end], next))
}
